simple-request-response/README.md

```diff
@@ -0,0 +1,75 @@
+# A simple request/response abstraction
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-07-14  |
+
+Authors: [@MarcoPolo]
+
+todo: others welcome to join
+Interest Group: [@mxinden @marten-seemann]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@MarcoPolo]: https://github.com/MarcoPolo
+[@mxinden]: https://github.com/mxinden
+
+## Introduction
+
+Many application protocols map well to request/response semantics where a peer
+makes a request to another peer, and receives a response back. For example,
+requesting a file from a peer and getting the file back. In contrast to
+request/response, stream semantics allow two peers to continously send data
+back and forth with no specific ordering. Both styles have their uses. This
+document focuses on _one_ way of implementing request/response semantics and
+how to map it to stream based transports and an HTTP transport.
+
+## Goals of this document
+
+A goal of this document is to define _a_ simple way to write request/response
+style application protocols that can make use of all libp2p transports
+seamlessly (all stream transports and the HTTP transport). This is only one way
+of writing request/response style protocols, and users may choose to not use it.
+
+Another goal is to be backwards compatible with existing application protocols
+that follow these semantics to make upgrading them easier so that they may take
+advantage of the new HTTP transport without sacrificing backwards compatibility.
+
+The final goal is to define something simple such that it's easy to follow,
+implement, and use.
+
+## The abstraction
+
+This abstraction takes in a blob from the requesting peer (aka client) that
+represents the request and delivers the blob to the responding peer (aka
+server). The server responds with a blob that respresents the response. This
+abstraction is agnostic to what the blobs are or how they are encoded. That's up
+to the application protocol to decide.
+
+## How to run on top of a libp2p stream
```

Comment:
My suggestion would to be to make this the main section of this document and define that this is how we do request-response on libp2p streams. I'd change the `SHOULD`s to `MUST` to make sure implementations can rely on that.

As mentioned above, I don't see how closing the write-side would be incompatible with any other way of doing request-response which is why I think it is safe to create this as a baseline abstraction.

Reply:
We could add here that implementations are free to use _additional_ ways of encoding the boundaries of req and res, for example by means of a TLV encoding.

Reply:
I think "MUST" would break backwards compatibility since existing clients may not close the write side after sending a request.

Reply:
> > plus it is easy to adapt existing implementations to additionally close the stream which makes them follow this spec.
> 
> Yes, it's easy to change new clients. But you can't change existing clients, thus breaking backwards compatibility.

Technically yes, but does it affect us? Is it important that our currently deployed protocols follow this spec? As far as I understand, we are mostly writing this to be able to deploy existing protocols over HTTP which requires code changes anyway. Can you outline a scenario where the incompatibility of existing clients with this spec hurts us?

> > each protocol has to define an additional mechanism like a TLV encoding to be able to reliably parse the messages
> 
> The application protocol themselves can say MUST. This doesn't prohibit that. If I make a new application protocol `/my-app/1` I can state in the spec that clients MUST close the stream when using the simple request-response abstraction.

Right but does that give us anything? If our goal is to write a generic piece of code, we cannot rely on what the individual protocols specify, right? In code terms, if a transports exposes an interface like:

```rust
async fn send_request(req: Res) -> Res;
```

How do you implement this if you cannot rely on receiving an EOF from the other party? If there is the possibility that a protocol does not say MUST, you effectively can't provide the above interface, otherwise it is not compatible with all possible protocols following this spec, right?

An implementation could be stricter and elevate the SHOULD to a MUST but that would limit which other protocols & implementations it is compatible with.